parsers/nsis.c

```diff
@@ -0,0 +1,249 @@
+/*
+*   Copyright (c) 2000-2002, Darren Hiebert
+*   Copyright (c) 2009-2011, Enrico Tröger
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License.
+*
+*   This module contains functions for generating tags for NSIS scripts
+*   (https://en.wikipedia.org/wiki/Nullsoft_Scriptable_Install_System).
+*
+*   Based on sh.c.
+*/
+
+/*
+*   INCLUDE FILES
+*/
+#include "general.h"	/* must always come first */
+
+#include <string.h>
+
+#include "entry.h"
+#include "parse.h"
+#include "read.h"
+#include "vstring.h"
+#include "routines.h"
+
+/*
+*   DATA DEFINITIONS
+*/
+typedef enum {
+	K_SECTION,
+	K_FUNCTION,
+	K_VARIABLE,
+	K_DEFINITION,
+	K_MACRO,
+	K_SECTION_GROUP,
+	K_MACRO_PARAM,
+} NsisKind;
+
+static kindDefinition NsisKinds [] = {
+	{ true, 's', "section", "sections"},
+	{ true, 'f', "function", "functions"},
+	{ true, 'v', "variable", "variables"},
+	{ true, 'd', "definition", "definitions"},
+	{ true, 'm', "macro", "macros"},
+	{ true, 'S', "sectionGroup", "section groups"},
+	{ false, 'p', "macroparam", "macro parameters"},
+};
+
+/*
+*   FUNCTION DEFINITIONS
+*/
+
+static const unsigned char* skipWhitespace (const unsigned char* cp)
+{
+	while (isspace ((int) *cp))
+		++cp;
+	return cp;
+}
+
+static const unsigned char* skipFlags (const unsigned char* cp)
+{
+	while (*cp == '/')
+	{
+		++cp;
+		while (! isspace ((int) *cp))
+			++cp;
+		while (isspace ((int) *cp))
+			++cp;
+	}
+	return cp;
+}
+
+static int makeSimpleTagWithScope(vString *name, int kindIndex, int parentCorkIndex)
+{
+	tagEntryInfo e;
+	initTagEntry (&e, vStringValue (name), kindIndex);
+	e.extensionFields.scopeIndex = parentCorkIndex;
+	return makeTagEntry (&e);
+}
+
+#define lineStartingWith(CP,EXPECTED,EOL)								\
+	(strncasecmp ((const char*) CP, EXPECTED, strlen(EXPECTED)) == 0	\
+		&& (EOL ? (isspace ((int) CP [strlen(EXPECTED)]) || CP [strlen(EXPECTED)] == '\0') \
+			: isspace ((int) CP [strlen(EXPECTED)])))
+
+#define fillName(NAME,CP,CONDITION)				\
+	while (CONDITION)							\
+	{											\
+		vStringPut ((NAME), (int) *(CP));		\
+		++(CP);									\
+	}											\
+	do {} while (0)
+
+static const unsigned char* parseSection (const unsigned char* cp, vString *name,
+										  int kindIndex, int scopeIndex, int *corkIndex)
+{
+	bool in_quotes = false;
+	cp = skipWhitespace (cp);
+	cp = skipFlags (cp);
+	while (isalnum ((int) *cp) || isspace ((int) *cp) ||
+		   *cp == '_' || *cp == '-' || *cp == '.' || *cp == '!' || *cp == '"'
+		   || (in_quotes && (*cp == '$' || *cp == '{' || *cp == '}' )))
```

Comment:
As far as I know, any characters can be used for a section(group) name when they are quoted.
Even Japanese characters can be used.